### PR TITLE
vala: Custom header and vapi name (fix #892)

### LINKED ISSUE #892: Allow installation of output header (-H) for Vala library targets
Currently it is unconditionally named after the library itself, but this is not always what people want. For instance in #890, @grindhold names the library `libhttpseverywhere-0.2`, but the header is called `libhttpseverywhere.h`.

Found while investigating #890

### PR DESCRIPTION
**Pull request by arteymix**
The same is done for the `--gir` flag in #634.

### CODE REVIEW

**Comment by nirbheek** [comment]
This should go in after #908 since it touches that code.

**Comment by arteymix** [comment]
@nirbheek  If you can cherry-pick and make it suitable for 0.36 that would be really nice.

Otherwise I think it's good to keep this open if Jussi decide to release before you're done.

**Comment by nirbheek** [comment]
Yes, of course we should keep this open. :) 

The patch needs some more work too; just that it's a bit pointless to do that right now because it's based on a branch that doesn't have necessary fixes yet. Will do that once that's done.

**Comment by arteymix** [comment]
@nirbheek what's the state of this with #908 merged?

**Comment by arteymix** [comment]
@nirbheek Ok, I just rebased upon your changes

Thanks for moving the VAPI out of `Executable` target, it's much cleaner now.

File: mesonbuild/build.py [hunk 1]
```diff
@@ -449,6 +451,9 @@ def process_kwargs(self, kwargs, environment):
         if not isinstance(valalist, list):
             valalist = [valalist]
         self.add_compiler_args('vala', valalist)
+        if not isinstance(self, Executable):
+            self.vala_header = kwargs.get('vala_header', None)
+            self.vala_vapi = kwargs.get('vala_vapi', None)
```
Comment (nirbheek) [inline mesonbuild/build.py]: Why not use `kwargs.get('vala_header', self.name + '.h')`, etc here? Simplifies the code a bit.

**Comment by arteymix** [comment]
I just merged the two test cases into one.

I'm not against introducing `install_vala_vapi` and `install_vala_header` options, although I feel like it's more appropriate to reuse existing concepts here (e.g `install_headers` and `install_data`).

File: mesonbuild/backend/ninjabackend.py [hunk 2]
```diff
@@ -992,15 +992,14 @@ def generate_vala_compile(self, target, outfile):
...
             args += ['-H', hname]
             valac_outputs.append(hname)
             # Outputted vapi file
-            base_vapi = target.name + '.vapi'
-            vapiname = os.path.join(self.get_target_dir(target), base_vapi)
+            vapiname = os.path.join(self.get_target_dir(target), target.vala_vapi)
             # Force valac to write the vapi file in the target build dir.
             # Without this, it will write it inside c_out_dir
-            args += ['--vapi=../' + base_vapi]
+            args += ['--vapi=../' + target.vala_vapi]
```
Comment (arteymix) [inline mesonbuild/backend/ninjabackend.py]: @nirbheek what's wrong with `get_target_dir` at this point? Isin't the C header generated in the same folder already?

Comment (arteymix) [inline mesonbuild/backend/ninjabackend.py]: Also, we should use `os.path.join` anyway for portability.

Comment (nirbheek) [inline mesonbuild/backend/ninjabackend.py]: I don't understand your question. We could use `os.path.join` here, sure.

Comment (arteymix) [inline mesonbuild/backend/ninjabackend.py]: There's two different strategy used to generate a file in what appears to be the same directory (`get_target_dir` vs `..`). It seems to me that the former is more correct, no?

Comment (arteymix) [inline mesonbuild/backend/ninjabackend.py]: If it's equivalent, I can patch it to consistently use `os.path.join` with `get_target_dir`.

Comment (nirbheek) [inline mesonbuild/backend/ninjabackend.py]: The two are not the same. `get_target_dir()` will return the subdir relative to the build root where the target is being built (the "target build dir"), which is not `..`.

For some reason, `valac` does not put the generated `.h` and the `.vapi` in the same directory. The `.h` file is placed in the directory as specified by the path, but the `.vapi` is placed alongside the generated C files in the 'target private directory' (see `c_out_dir` above) which is a subdir of the target build dir. The `..` is a hack to make it output the `.vapi` in the target build dir.

Try it out and see. You can make it `os.path.join('..', target.vala_vapi)` if you wish. :)

Comment (arteymix) [inline mesonbuild/backend/ninjabackend.py]: Then I need to check if the same apply for `vala_gir`.

**Pull request merged**